apps/labs/posts/libsf-error-state.md

```diff
@@ -0,0 +1,744 @@
+`sf_error_actions`.
+
+```c
+typedef enum {
+    SF_ERROR_OK = 0,    /* no error */
+    SF_ERROR_SINGULAR,  /* singularity encountered */
+    SF_ERROR_UNDERFLOW, /* floating point underflow */
+    SF_ERROR_OVERFLOW,  /* floating point overflow */
+    SF_ERROR_SLOW,      /* too many iterations required */
+    SF_ERROR_LOSS,      /* loss of precision */
+    SF_ERROR_NO_RESULT, /* no result obtained */
+    SF_ERROR_DOMAIN,    /* out of domain */
+    SF_ERROR_ARG,       /* invalid input parameter */
+    SF_ERROR_OTHER,     /* unclassified error */
+    SF_ERROR__LAST
+} sf_error_t;
+```
+
+`sf_action_t` is declared static to make use of its name local only to the file where it's defined. To work with this
+array externally, there are two functions, `scipy_sf_error_set_action` for updating and `scipy_sf_error_get_action` for
+querying the error handling policy associated to a given error type.
+
+```c
+void scipy_sf_error_set_action(sf_error_t code, sf_action_t action)
+{
+	sf_error_actions[(int)code] = action;
+}
+
+sf_action_t scipy_sf_error_get_action(sf_error_t code)
+{
+    return sf_error_actions[(int)code];
+}
+```
+
+If a user runs the following in an IPython session
+
+```python
+In  [1]: import scipy.special as special
+
+In  [2]: special.seterr(loss="warn")
+```
+
+then somewhere down in the stack, the following call will occur to update the error state.
+
+```c
+scipy_sf_error_set_action(SF_ERROR_LOSS, SF_ERROR_WARN)
+```
+
+If `special.gamma` is evaluated at an array containing a negative integer, then when the scalar kernel is evaluating the
+offending entry, it will signal `SF_ERROR_SINGULAR` and `sf_error_actions` will be consulted to figure out what action
+to take.
+
+## The many battles to actually ship it
+
+The shared library `libsf_error_state`'s contents are fairly simple, but how does one actually ship a shared library
+within a Python package like SciPy? The process for creating and using shared libraries depends on the operating system
+and compiler toolchain. SciPy supports a wide range of platforms in aim of its goal to democratize access to scientific
+computing tools and the greatest challenge turned out to be getting things to work on each of them. Several times, just
+as we thought everything was finally working, another quirk would pop up that needed to be addressed.
+
+### Path troubles
+
+The initial challenge was finding the right invocations to give to the meson build system used by SciPy []. Extension
+modules are configured in the `meson.build` files spread across SciPy's source tree and we needed to figure out how to
+set up a shared library and link it into each of the special function ufunc extension modules. Irwin and I begin working
+on this independently, comparing notes as we went.
+
+The first hiccup is that the following invocations were working on Irwin's Mac.
+
+Setting up the shared library like this.
+
+```meson
```

Comment:
Seems like we're missing pygments support for Meson here. A little unfortunate, it works on GitHub I believe. May want to change to marking this as python code if it can't work as meson syntax.